File: backend/app/api/v1/reports.py
from __future__ import annotations

from pathlib import Path

from fastapi import APIRouter, Depends
from fastapi.responses import FileResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.exceptions import NotFoundError
from app.schemas.reports import ReportRunCreate, ReportRunRead, ReportTemplateCreate, ReportTemplateRead
from app.services.report_service import ReportService

router = APIRouter(tags=["reports"])


@router.post("/report-templates", response_model=ReportTemplateRead, status_code=201)
async def create_report_template(
    payload: ReportTemplateCreate, session: AsyncSession = Depends(get_db)
):
    service = ReportService(session)
    template = await service.create_template(payload)
    return ReportTemplateRead.model_validate(template)


@router.post("/studies/{study_id}/reports", response_model=ReportRunRead, status_code=201)
async def generate_report(
    study_id: int, payload: ReportRunCreate, session: AsyncSession = Depends(get_db)
):
    service = ReportService(session)
    report = await service.generate(study_id, payload)
    return ReportRunRead.model_validate(report)


@router.get("/reports/{report_id}", response_model=ReportRunRead)
async def get_report(report_id: int, session: AsyncSession = Depends(get_db)):
    service = ReportService(session)
    report = await service.get_run(report_id)
    return ReportRunRead.model_validate(report)


@router.get("/reports/{report_id}/download")
async def download_report(report_id: int, session: AsyncSession = Depends(get_db)):
    service = ReportService(session)
    report = await service.get_run(report_id)
    if report.status != "COMPLETED" or not report.storage_path:
        raise NotFoundError(f"Reporte {report_id} no tiene un archivo disponible todavía")

    path = Path(report.storage_path)
    if not path.exists():
        raise NotFoundError(f"Archivo de reporte {report_id} no encontrado en almacenamiento")

    media_type = (
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        if path.suffix == ".docx"
        else "application/pdf"
        if path.suffix == ".pdf"
        else "application/json"
    )
    return FileResponse(path=path, media_type=media_type, filename=path.name)

File: frontend/src/api/reports.js
import { API_BASE_URL, apiRequest } from './client.js';

export function createReportTemplate(payload) {
  return apiRequest('/report-templates', { method: 'POST', body: payload });
}

export function createReport(studyId, payload) {
  return apiRequest(`/studies/${studyId}/reports`, { method: 'POST', body: payload });
}

export function createReportPreview(studyId, payload) {
  return apiRequest(`/studies/${studyId}/reports/preview`, { method: 'POST', body: payload });
}

export function getReport(reportId) {
  return apiRequest(`/reports/${reportId}`);
}

export function getReportDownloadUrl(reportId) {
  return `${API_BASE_URL}/reports/${reportId}/download`;
}

export function getReportPreviewPdfUrl(reportId) {
  return `${API_BASE_URL}/reports/${reportId}/preview.pdf`;
}
